### DISCUSSION TITLE
How to update json filed?

### DISCUSSION THREAD

**Discussion by heckad** [sqlalchemy/sqlalchemy#7033] [Usage Questions] [answered]
For exmaple I have a json `{'a':1, 'b':2}` in data json column and `{'b':3.'c':4}` as new date. I want to create update  expression wich after execute updated value in column to `{'a':1, 'b':3, 'c':4}`. How to do it using table clauses?

**Comment by zzzeek** [comment, marked as answer]
hey there -

first off, the easy way to get the new data into this column is to fully replace the JSON value with the new one that has what you want.  I'm assuming you know this already and you are looking for a means of manipulating just a small part of an existing JSON field without restating the full value.

first assumption is that you are on PostgreSQL.   if not, I dont think there are any options for in-place update of invididual JSON elements without updating the entire value.

For PG, while we support an UPDATE of specific fields inside the ARRAY datatype, a quick google doesn't show a PG operator that does this for JSON, what is offered is some manipulation functions and operators offered for the JSONB datatype only: https://stackoverflow.com/questions/18209625/how-do-i-modify-fields-inside-the-new-postgresql-json-datatype  

The first part where it shows "merging" fields seems most appropriate, and it's using the string concatenation operator "||".  SQLAlchemy offers the concat operator via the "+" sign or the ".concat()" function.

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import literal
from sqlalchemy import select
from sqlalchemy import update
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(JSONB)

e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

a1 = A(data={'a':1, 'b':2})
s.add(a1)
s.commit()

s.execute(
    update(A).values(data=A.data + literal({"b":3, "c": 4}, JSONB)),
    # or use concat function directly
    # update(A).values(data=A.data.concat(literal({"b":3, "c": 4}, JSONB)))
)

assert s.scalar(select(A.data)) == {"a": 1, "b": 3, "c": 4}
```

**Reply by heckad** [reply]
What to do if column can be nullable? In this case data will contain Null.

**Reply by CaselIT** [reply]
You would need to look at pg documentation. 
In general i think you can use `case` or `coalesce`
